lib/Sema/MiscDiagnostics.cpp

```diff
@@ -3455,15 +3458,32 @@ static void diagnoseUnintendedOptionalBehavior(TypeChecker &TC, const Expr *E,
       size_t optionalityDifference = 0;
       if (!isOptionalToAnyCoercion(srcType, destType, optionalityDifference))
         return;
-
-      TC.diagnose(subExpr->getStartLoc(), diag::optional_to_any_coercion,
-                  /* from */ srcType, /* to */ destType)
-        .highlight(subExpr->getSourceRange());
-
+      
+      // If we're implicitly unwrapping from IUO to Any then emit a custom
+      // diagnostic
+      if (hasImplicitlyUnwrappedResult(subExpr)) {
+        if (auto decl = getDeclForImplicitlyUnwrappedExpr(subExpr)) {
+          TC.diagnose(subExpr->getStartLoc(), diag::iuo_to_any_coercion,
+                      /* from */ srcType, /* to */ destType)
+              .highlight(subExpr->getSourceRange());
+
+          auto noteDiag = isa<FuncDecl>(decl)
+                              ? diag::iuo_to_any_coercion_note_func_result
+                              : diag::iuo_to_any_coercion_note;
+
+          TC.diagnose(decl->getLoc(), noteDiag, decl->getDescriptiveKind(),
+                      DeclName(decl->getBaseName()));
```

Comment:
How about `decl->getFullName()` ?

Reply:
Would that include the arg name as well?

Reply:
Yes, that would include labels.

Reply:
Yeah, which is why I didn’t use it

Reply:
Why not?

Reply:
Hmm I think it would be cleaner, but after thinking more about it, I guess it doesn't hurt to add arg labels as well. I have now made the change!